File: src/aoc/2015/day-1/main.rb
def get_basement_direction_position lines
  floor_number = 0
  for line in lines
    line.split("").each.with_index do |char, index|
      floor_number += get_direction_char_value char
      if floor_number == -1
        return index + 1
      end
    end
  end
  raise "Unable to Find Basement Direction Position"
end

def get_file_lines file_path
  IO.readlines(file_path, chomp: true)
end

def get_floor_number_from_directions lines
  floor_number = 0
  for line in lines
    for char in line.split("")
      floor_number += get_direction_char_value char
    end
  end
  floor_number
end

def get_direction_char_value char
  char == "(" ? 1 : char == ")" ? -1 : 0
end

INPUT_FILE_NAME = "./input.txt"

directions = get_file_lines INPUT_FILE_NAME
floor_number = get_floor_number_from_directions directions
position = get_basement_direction_position directions
puts "Floor Number: #{floor_number} | Basement Direction Position: #{position}"

File: src/aoc/2015/day-1/main.nim
import std/strformat

proc getDirectionCharValue(letter: char): int =
  if letter == '(':
    return 1
  elif letter == ')':
    return -1
  return 0

proc getBasementDirectionPosition(lines: seq[string]): int =
  var floorNumber = 0
  for line in lines:
    for index, letter in line:
      floorNumber += getDirectionCharValue(letter)
      if floorNumber == -1:
        return index + 1
  raise newException(ValueError, "Unable to Find Basement Direction Position")

proc getFllorNumberFromDirections(lines: seq[string]): int =
  var floorNumber = 0
  for line in lines:
    for letter in line:
      floorNumber += getDirectionCharValue(letter)
  return floorNumber

proc getFileLines(filePath: string): seq[string] =
  return io.readLines(filePath, 1)
  
const input_file_name = "./input.txt"

let directions = getFileLines(input_file_name)
let floorNumber = getFllorNumberFromDirections(directions)
let position = getBasementDirectionPosition(directions)
echo(fmt"Floor Number: {floorNumber} | Basement Direction Position: {position}")
